torch/distributed/_composable/fsdp/_fsdp_state.py

```diff
@@ -253,3 +288,46 @@ def _get_module_fsdp_state(module: nn.Module) -> Optional[FSDPState]:
     if isinstance(state, FSDPState):
         return state
     return None
+
+
+def _register_group_forward_hooks(
+    modules: Sequence[nn.Module],
+    pre_hook: Callable,
+    post_hook: Callable,
+):
+    """
+    Registers group forward pre and post-hooks. The pre-hook runs upon the
+    first module pre-forward, and the post-hook runs upon the last. If at least
+    one module does not run forward, then the post-hook does not run.
+    """
+    modules_set = set(modules)
+    modules_to_run: Set[nn.Module] = set()
+
+    @functools.wraps(pre_hook)
+    def wrapped_pre_hook(*args: Any, **kwargs: Any):
+        if len(modules_to_run) == 0:  # first to run
+            modules_to_run.update(modules_set)
+            return pre_hook(*args, **kwargs)
+
+    def get_wrapped_post_hook(module: nn.Module):
+        @functools.wraps(post_hook)
+        def wrapped_post_hook(*args: Any, **kwargs: Any):
+            modules_to_run.discard(module)
+            if len(modules_to_run) == 0:
+                return post_hook(*args, **kwargs)
+
+        return wrapped_post_hook
+
+    pre_handles = [
+        module.register_forward_pre_hook(
+            wrapped_pre_hook, prepend=True, with_kwargs=True
+        )
+        for module in modules
+    ]
+    post_handles = [
+        module.register_forward_hook(
+            get_wrapped_post_hook(module), prepend=False, always_call=True
```

Comment:
cc: @yf225 Do you know if `torch.compile` can handle `always_call=True` today?

Reply:
Previous context is in https://github.com/pytorch/pytorch/pull/127786#discussion_r1624911414

Reply:
I believe it should work with `torch.compile`. Wonder is this path triggered for typical FSDP2 use cases currently? If https://github.com/pytorch/pytorch/blob/973037be6a33c9cf114ac9172a564f4845636168/test/distributed/_composable/fsdp/test_fully_shard_compile.py#L300 passes then we are also in good shape for compile (seems like CI ran through so it's probably already the case)

Reply:
Ah, this code path is new, so none of the existing compile tests should exercise it. I can do a quick test to see if trying to use this code path with your compile tests works or not.